pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +223,238 @@ pub mod pallet {
+			}
+
+			Pools::<T>::remove(&pool_id);
+
+			// TODO: refund deposit
+
+			Self::deposit_event(Event::Destroyed(pool_id));
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_refund(pool_id: T::PoolId, maybe_check_manager: Option<&AccountIdOf<T>>) -> DispatchResult {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			// refunding can only be triggered on a live pool
+			ensure!(pool_details.state.is_live(), Error::<T>::PoolNotLive);
+
+			if let Some(caller) = maybe_check_manager {
+				ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized);
+			}
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+			// nothing to distribute
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::ZeroAmount);
+
+			let has_holders = pool_details
+				.bonded_currencies
+				.iter()
+				.any(|asset_id| !T::Fungibles::total_issuance(asset_id.clone()).is_zero());
+			// no token holders to refund
+			ensure!(has_holders, Error::<T>::ZeroAmount);
+
+			// move pool state to refunding
+			pool_details.state.refunding();
+			Pools::<T>::set(&pool_id, Some(pool_details));
+
+			Self::deposit_event(Event::RefundingStarted(pool_id));
+
+			Ok(())
+		}
+
+		fn do_start_destroy_pool(
+			pool_id: T::PoolId,
+			maybe_check_manager: Option<&AccountIdOf<T>>,
+			force_skip_refund: Fortitude, // TODO: enum or boolean flag?
+		) -> DispatchResult {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(
+				pool_details.state.is_live() || pool_details.state.is_refunding(),
+				Error::<T>::PoolNotLive
+			);
+
+			if let Some(caller) = maybe_check_manager {
+				ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized); // TODO: should this be permissionless if the pool is in refunding state?
+			}
+
+			if force_skip_refund != Fortitude::Force {
+				let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+				if !total_collateral_issuance.is_zero() {
+					let has_holders = pool_details
+						.bonded_currencies
+						.iter()
+						.any(|asset_id| !T::Fungibles::total_issuance(asset_id.clone()).is_zero());
+					// destruction is only allowed when there are no holders or no collateral to distribute
+					ensure!(!has_holders, Error::<T>::LivePool);
+				}
+			}
+
+			// move to destroying state
+			pool_details.state.destroy();
+			Pools::<T>::set(&pool_id, Some(pool_details.clone()));
+
+			// emit this event before the destruction started events are emitted by assets deactivation
+			Self::deposit_event(Event::DestructionStarted(pool_id));
+
+			// deactivate all currencies
+			for asset_id in pool_details.bonded_currencies.iter() {
+				// Governance or other pallets using the fungibles trait can in theory destroy an asset without this pallet knowing, so we check if it's still around
+				if T::Fungibles::asset_exists(asset_id.clone()) {
```

Comment:
So there's no `else`? What happens if the currency is destroyed elsewhere? Isn't that very bad for this curve?

Reply:
No, it's a workable even if cumbersome path of action to destroy the currencies via the assets pallet instead. This is a permissionless call and can be done by anyone as soon as the pool owner (this pallet) has moved the asset class to destroying state.

Reply:
Ok, makes sense.